File: Cpp/15.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> ans;
        set<vector<int>> s;
        for(int i = 0; i < nums.size()-2; i++){
            int j = i+1;
            int k = nums.size()-1;
            while(j < k){
                if(nums[i]+ nums[j]+ nums[k] == 0){
                    if(s.find({nums[i], nums[j], nums[k]}) == s.end()){
                        ans.push_back({nums[i], nums[j], nums[k]});
                        s.insert({nums[i], nums[j], nums[k]});
                    }
                    j++;
                }else if(nums[i]+ nums[j]+ nums[k] < 0){
                    j++;
                }else{
                    k--;
                }
            }
        }
        return ans;
        
    }
};

File: Python/15.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        sol = []
        i = 0
        while i < len(nums)-2:
            l = i+1
            r = len(nums)-1
            while l < r:
                if nums[l]+nums[r]+nums[i] < 0:
                    l+=1
                elif nums[l]+nums[r]+nums[i] > 0:
                    r-=1
                else:
                    sol.append([nums[i], nums[l], nums[r]])
                    x = nums[l]
                    y = nums[r]
                    while l < r and nums[l] == x:
                        l+=1
                    while l < r and nums[r] == y:
                        r-=1
            prev = nums[i]
            while i < len(nums)-2 and nums[i] == prev:
                i+=1
        return sol
